experimental/iterators/lib/Dialect/Iterators/Transforms/DecomposeIteratorStates.cpp

```diff
@@ -0,0 +1,243 @@
+    ValueRange extractedValue =
+        fieldMapping.getConvertedValues(convertedState, index);
+
+    // Replace original op with extracted field value.
+    rewriter.replaceOp(op, extractedValue, adaptor.getResultMapping());
+    return success();
+  }
+};
+
+void iterators::populateDecomposeIteratorStatesPatterns(
+    TypeConverter &typeConverter, RewritePatternSet &patterns) {
+  patterns.add<
+      // clang-format off
+      DecomposeExtractValueOp,
+      DecomposeInsertValueOp,
+      DecomposeCreateStateOp
+      // clang-format on
+      >(typeConverter, patterns.getContext());
+}
+
+static std::optional<Value> buildCreateStateOp(OpBuilder &builder,
+                                               TypeConverter &typeConverter,
+                                               Type type, ValueRange inputs,
+                                               Location loc) {
+  auto stateType = type.dyn_cast<StateType>();
+  if (!stateType)
+    return {};
+
+  SmallVector<Value> operands;
+  operands.reserve(stateType.getFieldTypes().size());
+  ValueRange::iterator inputsIt = inputs.begin();
+  for (Type fieldType : stateType.getFieldTypes()) {
+    Value operand;
+    if (fieldType.isa<StateType>()) {
+      SmallVector<Type> nestedFieldTypes;
+      if (failed(typeConverter.convertType(fieldType, nestedFieldTypes)))
+        return {};
+      ValueRange nestedFields{inputsIt, inputsIt + nestedFieldTypes.size()};
+      std::optional<Value> createdState = buildCreateStateOp(
+          builder, typeConverter, fieldType, nestedFields, loc);
+      if (!createdState.has_value() || !createdState.value())
+        return {};
+      operand = createdState.value();
+    } else {
+      operand = *inputsIt;
+    }
+    operands.push_back(operand);
+  }
+  auto createStateOp = builder.create<CreateStateOp>(loc, type, operands);
+  assert(createStateOp->getNumResults() == 1);
+  return createStateOp->getResult(0);
+}
+
+std::optional<SmallVector<Value>>
+buildExtractValueOps(OpBuilder &builder, TypeConverter &typeConverter,
+                     TypeRange resultTypes, Value input, Location loc) {
+  auto stateType = input.getType().dyn_cast<StateType>();
+  if (!stateType)
+    return {};
+
+  SmallVector<Value> extractedValues;
+  for (auto [idx, fieldType] : llvm::enumerate(stateType.getFieldTypes())) {
+    Value extractedValue =
+        builder.create<ExtractValueOp>(loc, input, builder.getIndexAttr(idx));
+
+    // If the value isn't a nested state, we can take it as is.
+    if (!fieldType.isa<StateType>()) {
+      extractedValues.push_back(extractedValue);
+      continue;
+    }
+
+    // For states, we recurse to extract the nested fields.
+    SmallVector<Type> nestedResultTypes;
+    if (failed(typeConverter.convertType(extractedValue.getType(),
+                                         nestedResultTypes)))
+      return {};
+    std::optional<SmallVector<Value>> maybeResults = buildExtractValueOps(
```

Comment:
this is recursive but the depth is small ?

Reply:
Fixed in latest commit.